virtual_ship/instruments/adcp.py

```diff
@@ -0,0 +1,66 @@
+"""ADCP instrument."""
+
+import numpy as np
+from parcels import FieldSet, JITParticle, ParticleSet, Variable
+
+from ..spacetime import Spacetime
+
+_ADCPParticle = JITParticle.add_variables(
+    [
+        Variable("U", dtype=np.float32, initial=np.nan),
+        Variable("V", dtype=np.float32, initial=np.nan),
+    ]
+)
+
+
+def _sample_velocity(particle, fieldset, time):
+    particle.U, particle.V = fieldset.UV.eval(
+        time, particle.depth, particle.lat, particle.lon, applyConversion=False
+    )
+
+
+def simulate_adcp(
+    fieldset: FieldSet,
+    out_file_name: str,
+    max_depth: float,
+    min_depth: float,
+    bin_size: float,
+    sample_points: list[Spacetime],
+) -> None:
+    """
+    Use parcels to simulate an ADCP in a fieldset.
+
+    :param fieldset: The fieldset to simulate the ADCP in.
+    :param out_file_name: The file to write the results to.
+    :param max_depth: Maximum depth the ADCP can measure.
+    :param min_depth: Minimum depth the ADCP can measure.
+    :param bin_size: How many samples to take in the complete range between max_depth and min_depth.
+    :param sample_points: The places and times to sample at.
+    """
+    sample_points.sort(key=lambda p: p.time)
+
+    bins = np.arange(max_depth, min_depth, bin_size)
+    num_particles = len(bins)
+    particleset = ParticleSet.from_list(
+        fieldset=fieldset,
+        pclass=_ADCPParticle,
+        lon=np.full(
+            num_particles, 0.0
+        ),  # initial lat/lon are irrelevant and will be overruled later.
+        lat=np.full(num_particles, 0.0),
+        depth=bins,
+        time=0,  # same for time
+    )
+
+    # define output file for the simulation
+    out_file = particleset.ParticleFile(
+        name=out_file_name,
```

Comment:
I was already working on the follow up pr where I implement proper tests. Here, we explicitly only do one step during execute, always writing to the file each iteration. (not very efficient but its fine as we don't have many sampling points anyway). As long as the outputdt is larger than the step size we are fine. The default is float.infinite so that's good.